### PR TITLE
add an unknown-origin asset to add mappings that have no defined source

### PR DESCRIPTION
**Pull request by DeMoorJasper**
# ↪️ Pull Request

This PR rewrites the build report utility slightly to include an `unknown-origin` dummy asset to highlight mappings that have no source. This mainly happens when adding in additional data into an ast as these do not originate from a specific file.

This also solves the issue where some assets are larger than the complete bundle by normalising paths and making sure once it's a sourcemap based bundlereport we only care about the sources in the sourcemap and don't use sizes from the original assets as a reference.

This has been discovered by @mischnic in the ReactHN benchmark

With scope-hoisting

Screenshot 2020-05-29 at 13 42 08

Without scope-hoisting

Screenshot 2020-05-29 at 13 47 42

### CODE REVIEW

File: packages/core/utils/src/generateBuildMetrics.js [hunk 1]
```diff
@@ -85,10 +86,16 @@ async function getSourcemapSizes(
...
     let sizeMap = new Map();
+    for (let source of sources) {
+      sizeMap.set(source, 0);
+    }
+
     for (let i = 0; i < sourceSizes.length; i++) {
       sizeMap.set(sources[i], sourceSizes[i]);
     }
 
+    sizeMap.set('unknown-origin', unknownOrigin);
```
Comment (mischnic) [inline packages/core/utils/src/generateBuildMetrics.js:97]: Maybe do `sizeMap.set(null, unknownOrigin)` instead so that the reporter can do some special formatting for this pseudo asset? Right now it looks like a regular asset that is just called "unknown-origin".

Comment (DeMoorJasper) [reply packages/core/utils/src/generateBuildMetrics.js:97]: Sure, I'm just not sure how I want to display it.

**Comment by parcel-benchmark** [comment]
## Benchmark Results
<details><summary>Kitchen Sink ✅</summary><p>

#### Timings

| Description | Time | Difference |
| --- | --- | --- |
| Cold | 6.22s | -114.00ms |
| Cached | 2.26s | -7.00ms |

#### Cold Bundles

| Bundle | Size | Difference | Time | Difference |
| --- | --- | --- | --- | --- |
| dist/legacy/parcel.d5807e82.webp | 102.94kb | +0.00b | 113.00ms | -107.00ms 🚀 |
| dist/modern/parcel.d5807e82.webp | 102.94kb | +0.00b | 113.00ms | -108.00ms 🚀 |
| dist/legacy/kitchen-sink.679578a5.js | 1.16kb | +0.00b | 1.41s | -253.00ms 🚀 |
| dist/modern/kitchen-sink.90de237b.js | 1.16kb | +0.00b | 529.00ms | -1.14s 🚀 |
| dist/legacy/index.html | 709.00b | +0.00b | 648.00ms | -40.00ms 🚀 |
| dist/modern/index.html | 709.00b | +0.00b | 646.00ms | -40.00ms 🚀 |
| dist/legacy/styles.afb8e31a.css | 78.00b | +0.00b | 1.41s | -252.00ms 🚀 |

#### Cached Bundles

| Bundle | Size | Difference | Time | Difference |
| --- | --- | --- | --- | --- |
| dist/legacy/parcel.d5807e82.webp | 102.94kb | +0.00b | 192.00ms | +36.00ms ⚠️ |
| dist/modern/parcel.d5807e82.webp | 102.94kb | +0.00b | 193.00ms | +38.00ms ⚠️ |
| dist/legacy/kitchen-sink.679578a5.js | 1.16kb | +0.00b | 358.00ms | +33.00ms ⚠️ |
| dist/modern/kitchen-sink.90de237b.js | 1.16kb | +0.00b | 357.00ms | +33.00ms ⚠️ |
| dist/legacy/index.html | 709.00b | +0.00b | 359.00ms | +33.00ms ⚠️ |
| dist/modern/index.html | 709.00b | +0.00b | 357.00ms | +34.00ms ⚠️ |
| dist/legacy/styles.afb8e31a.css | 78.00b | +0.00b | 358.00ms | +34.00ms ⚠️ |
| dist/modern/styles.cda8c91b.css | 78.00b | +0.00b | 357.00ms | +33.00ms ⚠️ |

</p></details>

<details><summary>React HackerNews ✅</summary><p>

#### Timings

| Description | Time | Difference |
| --- | --- | --- |
| Cold | 13.85s | +492.00ms |
| Cached | 2.30s | -19.00ms |

#### Cold Bundles

| Bundle | Size | Difference | Time | Difference |
| --- | --- | --- | --- | --- |
| dist/main/index.js | 66.80kb | +0.00b | 2.18s | -116.00ms 🚀 |
| dist/module/index.js | 37.47kb | +0.00b | 2.17s | -115.00ms 🚀 |

#### Cached Bundles

| Bundle | Size | Difference | Time | Difference |
| --- | --- | --- | --- | --- |
| dist/main/index.js | 66.80kb | +0.00b | 377.00ms | +104.00ms ⚠️ |
| dist/module/index.js | 37.47kb | +0.00b | 376.00ms | +105.00ms ⚠️ |

</p></details>

<details><summary>AtlasKit Editor ✅</summary><p>

#### Timings

| Description | Time | Difference |
| --- | --- | --- |
| Cold | 2.78m | -11.44s 🚀 |
| Cached | 5.28s | +3.00ms |

#### Cold Bundles

| Bundle | Size | Difference | Time | Difference |
| --- | --- | --- | --- | --- |
| dist/ak-editor.74f6fc0d.js | 2.50mb | +4.00b ⚠️ | 1.58m | -9.31s 🚀 |
| dist/pdfRenderer.74642224.js | 1.11mb | -6.00b 🚀 | 57.53s | +4.46s ⚠️ |
| dist/editorView.54e7eb37.js | 586.92kb | +0.00b | 1.56m | +7.49s ⚠️ |
| dist/popup.ff1c0207.js | 218.72kb | +0.00b | 1.09m | +9.05s ⚠️ |
| dist/ak-editor.7933d476.js | 136.51kb | +0.00b | 17.46s | -1.44m 🚀 |
| dist/Toolbar.4e6aa6b0.js | 100.36kb | +0.00b | 1.11m | +6.77s ⚠️ |
| dist/EmojiPickerComponent.2c06cc58.js | 98.96kb | +0.00b | 17.46s | -1.44m 🚀 |
| dist/media-viewer.0b37a050.js | 71.13kb | +0.00b | 21.34s | +5.56s ⚠️ |
| dist/card.306e2acc.js | 66.70kb | +0.00b | 21.34s | +5.56s ⚠️ |
| dist/component.7a54e3cc.js | 55.63kb | +0.00b | 6.68s | -885.00ms 🚀 |
| dist/Modal.83b93036.js | 41.59kb | +0.00b | 3.69s | -583.00ms 🚀 |
| dist/DatePicker.d24458b4.js | 21.54kb | +0.00b | 9.08s | -668.00ms 🚀 |
| dist/EmojiPickerComponent.a2474c55.js | 18.34kb | +0.00b | 52.73s | +42.41s ⚠️ |
| dist/clipboard.2b70f60b.js | 17.18kb | +0.00b | 1.58m | +37.56s ⚠️ |
| dist/js.baeda348.js | 16.46kb | +0.00b | 6.61s | -880.00ms 🚀 |
| dist/dropzone.88563fa4.js | 14.69kb | +0.00b | 1.58m | +37.58s ⚠️ |
| dist/smartMediaEditor.ab4310d1.js | 14.62kb | +0.00b | 1.04m | +5.46s ⚠️ |
| dist/ui.9e4d74b0.js | 14.45kb | +0.00b | 21.34s | +5.58s ⚠️ |
| dist/workerHasher.ebf0b386.js | 11.83kb | +0.00b | 52.73s | -9.50s 🚀 |
| dist/EmojiTypeAheadComponent.899f7d25.js | 9.20kb | +0.00b | 17.46s | +4.99s ⚠️ |
| dist/media-viewer.52d4959e.js | 4.18kb | +0.00b | 34.95s | +18.97s ⚠️ |
| dist/EmojiPickerComponent.bd593806.js | 3.92kb | +0.00b | 52.74s | +42.41s ⚠️ |
| dist/16.cb2f1a3a.js | 3.72kb | +0.00b | 3.19s | -1.01s 🚀 |
| dist/Modal.1111ef83.js | 3.50kb | +0.00b | 6.44s | +2.23s ⚠️ |
| dist/png-chunks-extract.c0f57e67.js | 3.49kb | +0.00b | 6.61s | -882.00ms 🚀 |
| dist/dropzone.79e05ece.js | 3.44kb | +0.00b | 1.58m | +41.55s ⚠️ |
| dist/ak-editor.47799eae.css | 3.43kb | +0.00b | 1.32m | +18.20s ⚠️ |
| dist/16.c7be7376.js | 2.54kb | +0.00b | 6.61s | -879.00ms 🚀 |
| dist/EmojiUploadComponent.4946a1c6.js | 2.50kb | +0.00b | 17.46s | +7.13s ⚠️ |
| dist/card.c002a0cc.js | 2.19kb | +0.00b | 57.04s | +41.02s ⚠️ |
| dist/16.bb063a1f.js | 1.83kb | +0.00b | 6.88s | -871.00ms 🚀 |
| dist/media-picker-analytics-error-boundary.187d26a0.js | 1.82kb | +0.00b | 1.58m | +41.55s ⚠️ |
| dist/date.56d1aa9e.js | 1.81kb | +0.00b | 7.22s | -802.00ms 🚀 |
| dist/browser.90c1ebeb.js | 1.79kb | +0.00b | 1.04m | +5.49s ⚠️ |
| dist/images.16b3af40.js | 1.75kb | +0.00b | 7.29s | -801.00ms 🚀 |
| dist/feedback.b59078b6.js | 1.72kb | +0.00b | 8.13s | -1.20s 🚀 |
| dist/workerHasher.9975765a.js | 1.65kb | +0.00b | 15.98s | -44.69s 🚀 |
| dist/ak-editor.7db5aa7b.js | 1.60kb | +0.00b | 1.11m | +6.79s ⚠️ |
| dist/list-number.d2097959.js | 1.54kb | +0.00b | 7.42s | -781.00ms 🚀 |
| dist/status.f61d9ac5.js | 1.54kb | +0.00b | 7.63s | -711.00ms 🚀 |
| dist/16.450f3a0d.js | 1.49kb | +0.00b | 6.84s | -839.00ms 🚀 |
| dist/code.4289d7dd.js | 1.46kb | +0.00b | 7.13s | -874.00ms 🚀 |
| dist/16.7d4b2402.js | 1.42kb | +0.00b | 7.05s | -853.00ms 🚀 |
| dist/link.36b44af7.js | 1.38kb | +0.00b | 7.36s | -793.00ms 🚀 |
| dist/heading6.675c185b.js | 1.38kb | +0.00b | 8.13s | -1.20s 🚀 |
| dist/16.6b0e338c.js | 1.38kb | +0.00b | 7.05s | -852.00ms 🚀 |
| dist/heading3.4e88bcd6.js | 1.37kb | +0.00b | 7.80s | -618.00ms 🚀 |
| dist/16.e32aa49c.js | 1.34kb | +0.00b | 3.19s | -1.01s 🚀 |
| dist/16.50f9c790.js | 1.33kb | +0.00b | 6.98s | -892.00ms 🚀 |
| dist/emoji.3654aec7.js | 1.31kb | +0.00b | 7.29s | -795.00ms 🚀 |
| dist/media-card-analytics-error-boundary.9ca79f13.js | 1.31kb | +0.00b | 1.02m | +7.89s ⚠️ |
| dist/16.81746ce1.js | 1.29kb | +0.00b | 6.98s | -892.00ms 🚀 |
| dist/16.d1b58d7c.js | 1.29kb | +0.00b | 7.13s | -866.00ms 🚀 |
| dist/16.1dbb5df4.js | 1.28kb | +0.00b | 3.19s | -1.01s 🚀 |
| dist/heading5.e338f1e8.js | 1.25kb | +0.00b | 8.13s | -1.20s 🚀 |
| dist/16.ceac57cd.js | 1.25kb | +0.00b | 6.88s | -867.00ms 🚀 |
| dist/16.35be54a4.js | 1.23kb | +0.00b | 6.84s | -838.00ms 🚀 |
| dist/16.6be9bbc8.js | 1.23kb | +0.00b | 6.84s | -840.00ms 🚀 |
| dist/16.9015e8cf.js | 1.23kb | +0.00b | 6.62s | -930.00ms 🚀 |
| dist/16.234ee133.js | 1.23kb | +0.00b | 3.19s | -3.40s 🚀 |
| dist/heading2.89a8e126.js | 1.19kb | +0.00b | 7.70s | -705.00ms 🚀 |
| dist/16.7ca41ffe.js | 1.16kb | +0.00b | 6.98s | -894.00ms 🚀 |
| dist/mention.ae379620.js | 1.15kb | +0.00b | 7.42s | -787.00ms 🚀 |
| dist/heading4.b14d5ea4.js | 1.14kb | +0.00b | 7.80s | -631.00ms 🚀 |
| dist/layout.d9df99ef.js | 1.13kb | +0.00b | 7.36s | -791.00ms 🚀 |
| dist/divider.4476c3de.js | 1.11kb | +0.00b | 7.29s | -795.00ms 🚀 |
| dist/quote.4733acb8.js | 1.10kb | +0.00b | 7.63s | -698.00ms 🚀 |
| dist/action.b9409056.js | 1.08kb | +0.00b | 7.13s | -867.00ms 🚀 |
| dist/decision.77ee0aaa.js | 1.07kb | +0.00b | 7.22s | -810.00ms 🚀 |
| dist/panel-warning.e55309d2.js | 1.06kb | +0.00b | 7.56s | -749.00ms 🚀 |
| dist/list.cf500a90.js | 1.03kb | +0.00b | 7.42s | -781.00ms 🚀 |
| dist/heading1.165705a7.js | 1.03kb | +0.00b | 7.70s | -708.00ms 🚀 |
| dist/media-card-analytics-error-boundary.a014425d.js | 995.00b | +0.00b | 57.04s | +41.02s ⚠️ |
| dist/panel-error.364fbf3c.js | 987.00b | +0.00b | 7.47s | -769.00ms 🚀 |
| dist/panel.e4fe05d1.js | 978.00b | +0.00b | 7.56s | -765.00ms 🚀 |
| dist/table.fd1280b6.js | 970.00b | +0.00b | 7.70s | -703.00ms 🚀 |
| dist/panel-success.66723ab0.js | 926.00b | +0.00b | 7.56s | -749.00ms 🚀 |
| dist/panel-note.a8b696c4.js | 922.00b | +0.00b | 7.47s | -770.00ms 🚀 |
| dist/media-viewer-analytics-error-boundary.ee954f0e.js | 892.00b | +0.00b | 35.06s | +19.08s ⚠️ |
| dist/EmojiTypeAheadComponent.441c229c.js | 619.00b | +0.00b | 19.81s | +4.05s ⚠️ |
| dist/simpleHasher.dd9d4306.js | 609.00b | +0.00b | 52.72s | +42.40s ⚠️ |
| dist/index.html | 219.00b | +0.00b | 1.01s | -629.00ms 🚀 |

#### Cached Bundles

*No bundle changes detected.*

</p></details>

<details><summary>Three.js x4 🚨</summary><p>

#### Timings

| Description | Time | Difference |
| --- | --- | --- |
| Cold | **FAILED** | -0.00ms |
| Cached | **FAILED** | -0.00ms |

#### Cold Bundles

*No bundles found, this is probably a failed build...*
#### Cached Bundles

*No bundles found, this is probably a failed build...*
</p></details>

**[Click here to view a detailed benchmark overview.](https://benchmark.parceljs.org/benchmark/266858786966209032)**

File: packages/core/utils/src/generateBuildMetrics.js [hunk 2]
```diff
@@ -85,10 +86,16 @@ async function getSourcemapSizes(
     }
 
     let sizeMap = new Map();
+    for (let source of sources) {
+      sizeMap.set(source, 0);
```
Comment (mischnic) [inline packages/core/utils/src/generateBuildMetrics.js:90]: Why is this needed? `let sourceSizes = new Array(sources.length).fill(0)` should handle this already?

Comment (DeMoorJasper) [reply packages/core/utils/src/generateBuildMetrics.js:90]: Ow yeah...

File: packages/reporters/cli/src/bundleReport.js [hunk 3]
```diff
@@ -53,13 +53,20 @@ export default async function bundleReport(
...
+          asset == largestAssets[largestAssets.length - 1] ? '└── ' : '├── ',
           chalk.dim(prettifySize(asset.size)),
           chalk.dim(chalk.green(prettifyTime(asset.time))),
-        ]);
+        ];
+
+        if (asset.filePath !== '') {
+          columns[0] += formatFilename(asset.filePath, chalk.reset);
+        } else {
+          columns[0] += 'Code from unknown sourcefile';
```
Comment (mischnic) [inline packages/reporters/cli/src/bundleReport.js:65]:
```suggestion
          columns[0] += 'Code from unknown sourcefiles';
```
